src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java

```diff
@@ -1261,81 +1382,320 @@ public ContentBuilder add(CharSequence text) {
+         */
+        private static final char PLACEHOLDER_CHAR = '\uFFFC';
+        /**
+         * Placeholder for non-HTML nodes that are not phrasing content.
+         * The placeholder is an HTML block, according to CommonMark 4.6 rule 2,
+         * and so will not be wrapped into with {@code <p>...</p>} tags.
+         * @see <a href="https://spec.commonmark.org/0.30/#html-blocks">HTML blocks</a>
+         */
+        private static final String PLACEHOLDER_BLOCK = "<!--\uFFFC-->";
+        private static final Pattern PLACEHOLDERS = Pattern.compile(PLACEHOLDER_CHAR + "|" + PLACEHOLDER_BLOCK);
+
+        private final StringBuilder markdownInput = new StringBuilder() ;
+        private final ArrayList<Content> fffcObjects = new ArrayList<>();
+
+        private final Extension tablesExtn = TablesExtension.create();
+        private final HtmlNodeRendererFactory headingRendererFactory = HeadingNodeRenderer::new;
+
+        private final Element element;
+
+        private final Parser parser = Parser.builder()
+                .extensions(List.of(tablesExtn))
+                .build();
+        private final HtmlRenderer renderer = HtmlRenderer.builder()
+                .nodeRendererFactory(headingRendererFactory)
+                .extensions(List.of(tablesExtn/*, headingIdsExtn*/))
+                .build();
+
+        MarkdownHandler(Element element) {
+            this.element = element;
+        }
+
+        boolean handle(DocTree tree, InlineVisitor visitor) {
+            boolean allDone;
+            if (tree instanceof RawTextTree t) {
+                if (t.getKind() != Kind.MARKDOWN) {
+                    throw new IllegalStateException(t.getKind().toString());
+                }
+                String code = t.getContent();
+                // handle the (unlikely) case of FFFC characters existing in the code
+                int start = 0;
+                int pos;
+                while ((pos = code.indexOf(PLACEHOLDER_CHAR, start)) != -1) {
+                    markdownInput.append(code.substring(start, pos));
+                    markdownInput.append(PLACEHOLDER_CHAR);
+                    fffcObjects.add(Text.of(String.valueOf(PLACEHOLDER_CHAR)));
+                    start = pos + 1;
                 }
+                markdownInput.append(code.substring(start));
+                allDone = false;
+            } else {
+                Content embeddedContent = new ContentBuilder();
+                allDone = visitor.visit(tree, embeddedContent);
+                fffcObjects.add(embeddedContent);
+                if (embeddedContent.isPhrasingContent()) {
+                    markdownInput.append(PLACEHOLDER_CHAR);
+                } else {
+                    if (!markdownInput.isEmpty() && markdownInput.charAt(markdownInput.length() - 1) != '\n') {
+                        markdownInput.append('\n');
+                    }
+                    markdownInput.append(PLACEHOLDER_BLOCK);
```

Comment:
There's quite a lot of overlap with the recently refactored https://github.com/openjdk/jdk/blob/7c6316886d1ae86a663d996dae373c42281622fd/src/jdk.internal.md/share/classes/jdk/internal/markdown/MarkdownTransformer.java#L220-L238

If it's impractical to factor out the common bits, maybe we could at least make the respective parts of HtmlDocletWriter as close as possible to those of MarkdownTransformer.

Reply:
I think it is impractical to factor out the common bits, but I will look at the possibility of making the code more similar -- but no promises.